### PR TITLE
Ch 15

### PR DESCRIPTION
**Pull request by devKate**
hello nish :D

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -4,13 +4,14 @@
...
     <application
+        android:name=".ui.App"
         android:allowBackup="true"
         android:icon="@mipmap/ic_launcher"
         android:label="@string/app_name"
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
         android:theme="@style/AppTheme">
-        <activity android:name=".Activities.MainActivity">
+        <activity android:name="com.bignerdranch.android.weatherapp.ui.Activities.MainActivity">
```
Comment (nishtahir) [inline app/src/main/AndroidManifest.xml]: Fully qualified class names are optional.

File: app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt [hunk 2]
```diff
@@ -20,7 +25,6 @@ class Request(private val zip: String) {
 //        val response: ForecastResponse = Gson().fromJson(tempResponseJSON, ForecastResponse::class.java)
```
Comment (nishtahir) [inline app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt]: Don't leave commented out code in your project

File: app/src/main/java/com/bignerdranch/android/weatherapp/ui/Utilities/DelegatesExt.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package com.bignerdranch.android.weatherapp.ui.Utilities
+
+import kotlin.reflect.KProperty
+
+/**
+ * Created by citylights on 8/29/17.
+ */
+
+class NonNullSingleValueVar<T> {
```
Comment (nishtahir) [inline app/src/main/java/com/bignerdranch/android/weatherapp/ui/Utilities/DelegatesExt.kt]: Looks like this is a fancy delegate that holds an optional value. It may have just been an excuse to show you how kotlin reflection works.

File: app/src/main/java/com/bignerdranch/android/weatherapp/ui/App.kt [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package com.bignerdranch.android.weatherapp.ui
+
+import android.app.Application
+import com.bignerdranch.android.weatherapp.ui.Utilities.NonNullSingleValueVar
+
+class App : Application() {
+    companion object {
+        var instance: App by NonNullSingleValueVar()
```
Comment (nishtahir) [inline app/src/main/java/com/bignerdranch/android/weatherapp/ui/App.kt]: I'd prefer dagger over holding a companion object reference to your application. for the most part you can use. 

```
context.getApplicationContext()
```
if you want a reference to your app context.

File: app/src/main/java/com/bignerdranch/android/weatherapp/Domain/Commands/RequestForecastCommand.kt [hunk 5]
```diff
@@ -9,7 +9,13 @@ import com.bignerdranch.android.weatherapp.Domain.Model.ForecastModel
...
  * Created by citylights on 8/27/17.
  */
-class RequestForecastCommand(val zip: String): Command<ForecastListModel> {
+
+/*
+* implement a command which takes a zipcode
+* makes an API request for a forecast
+* and returns a ForecastListModel
+* */
+class RequestForecastCommand(private val zip: String): Command<ForecastListModel> {
```
Comment (nishtahir) [inline app/src/main/java/com/bignerdranch/android/weatherapp/Domain/Commands/RequestForecastCommand.kt]: Shouldn't everything be private by default?

File: app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt [hunk 6]
```diff
@@ -6,6 +6,11 @@ import java.net.URL
 /**
  * Created by citylights on 8/26/17.
  */
+
+/*
```
Comment (nishtahir) [inline app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt]: Not sure why this was broken out into 2 multiline comments

File: app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt [hunk 7]
```diff
@@ -6,6 +6,11 @@ import java.net.URL
...
+* Takes a zipcode and makes an api request for the weather in that zipcode
+* parses the api's json response into raw response data (not the same as our business logic forecast models)
+* */
 class Request(private val zip: String) {
     companion object {
         private val APP_ID = "15646a06818f61f7b8d7823ca833e1ce"
```
Comment (nishtahir) [inline app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt]: You don't want to keep this here. since it could be potentially changed. I like to keep this in my `build.gradle`

File: app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt [hunk 8]
```diff
@@ -20,7 +25,6 @@ class Request(private val zip: String) {
 //        val response: ForecastResponse = Gson().fromJson(tempResponseJSON, ForecastResponse::class.java)
 //        return response
         return Gson().fromJson(forecastJSON, ForecastResponse::class.java)
```
Comment (nishtahir) [inline app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt]: You shouldn't need more than 1 gson in the app. You probably want to inject this.

File: app/src/main/java/com/bignerdranch/android/weatherapp/Domain/Commands/Command.kt [hunk 9]
```diff
@@ -3,7 +3,8 @@ package com.bignerdranch.android.weatherapp.Domain.Commands
 /**
  * Created by citylights on 8/26/17.
  */
-// execute an operation, return object of type T (generic)
-public interface Command<out T> {
+// execute an operation, return object of type T
+// @NISH is this pattern normal? why do we need the interface?
+interface Command<out T> {
```
Comment (nishtahir) [inline app/src/main/java/com/bignerdranch/android/weatherapp/Domain/Commands/Command.kt]: Looks like your architecture requires this.

File: app/src/main/java/com/bignerdranch/android/weatherapp/Domain/Commands/RequestForecastCommand.kt [hunk 10]
```diff
@@ -9,7 +9,13 @@ import com.bignerdranch.android.weatherapp.Domain.Model.ForecastModel
...
     override fun execute(): ForecastListModel {
         val forecastResponse = Request(zip).execute()
         return ForecastDataMapper().forecastResponseToForecastList(forecastResponse)
```
Comment (nishtahir) [inline app/src/main/java/com/bignerdranch/android/weatherapp/Domain/Commands/RequestForecastCommand.kt]: You're creating a new instance every time you run this. Probably want to inject this too.

File: app/src/main/java/com/bignerdranch/android/weatherapp/ui/Adapters/ForecastListAdapter.kt [hunk 11]
```diff
@@ -1,24 +1,21 @@
...
 import kotlinx.android.synthetic.main.item_forecast.view.*
 
 /**
  * Created by citylights on 8/26/17.
  */
-fun ImageView.loadImage(url: String) {
-    Picasso.with(context).load(url).into(this)
-}
 
 class ForecastListAdapter(private val weekForecastList: ForecastListModel, val itemClick: (ForecastModel) -> Unit) : RecyclerView.Adapter<ForecastListAdapter.ForecastViewHolder>() {
```
Comment (nishtahir) [inline app/src/main/java/com/bignerdranch/android/weatherapp/ui/Adapters/ForecastListAdapter.kt]: Not sure why inside your view holder you are referencing the `itemView`. Can't you reference the elements inside of the view holder directly? If you can't, you always have the option of doing `findViewById` inside of an `init` blocks.

File: app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt (hunk 2 above)
Comment (devKate) [reply app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt]: This is just temporary because the emulator wouldn't connect to the internet while I was behind a proxy at starbucks and I thought it would be easier than trying to configure the thing

File: app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt (hunk 8 above)
Comment (devKate) [reply app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt]: Where should I instantiate it?

File: app/src/main/java/com/bignerdranch/android/weatherapp/Domain/Commands/Command.kt (hunk 9 above)
Comment (devKate) [reply app/src/main/java/com/bignerdranch/android/weatherapp/Domain/Commands/Command.kt]: but, couldn't I just define the RequestForecastCommand without this Command interface? What is its purpose?

File: app/src/main/java/com/bignerdranch/android/weatherapp/Domain/Commands/RequestForecastCommand.kt (hunk 5 above)
Comment (devKate) [reply app/src/main/java/com/bignerdranch/android/weatherapp/Domain/Commands/RequestForecastCommand.kt]: the IDE was telling me to mark that as private, I thought it was public by default

File: app/src/main/java/com/bignerdranch/android/weatherapp/ui/Utilities/DelegatesExt.kt (hunk 3 above)
Comment (devKate) [reply app/src/main/java/com/bignerdranch/android/weatherapp/ui/Utilities/DelegatesExt.kt]: yes, they were saying you'd just use `lateinit` with a private setter in the real world

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (devKate) [reply app/src/main/AndroidManifest.xml]: changed this :D

File: app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt (hunk 6 above)
Comment (devKate) [reply app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt]: the top comment was generated by the IDE so I didn't want to mess with it. I could if that is better style but I sort of considered it something that should be the same on all files

File: app/src/main/java/com/bignerdranch/android/weatherapp/ui/Adapters/ForecastListAdapter.kt (hunk 11 above)
Comment (devKate) [reply app/src/main/java/com/bignerdranch/android/weatherapp/ui/Adapters/ForecastListAdapter.kt]: is it bad to reference the itemView there? I was thinking that that is how you are supposed to access subviews for each recyclerview list item

File: app/src/main/java/com/bignerdranch/android/weatherapp/ui/App.kt (hunk 4 above)
Comment (devKate) [reply app/src/main/java/com/bignerdranch/android/weatherapp/ui/App.kt]: 1. does context.getApplicationContext() return the custom application singleton we define here?
2. are there different contexts? what is the difference between context in an activity and this ApplicationContext?

File: app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt (hunk 2 above)
Comment (nishtahir) [reply app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt]: It's better to rely on build configs and build flavors, than commented out code in the repo. Comments that add no value build up very fast and make the code base messy and hard to follow.

File: app/src/main/java/com/bignerdranch/android/weatherapp/ui/App.kt (hunk 4 above)
Comment (nishtahir) [reply app/src/main/java/com/bignerdranch/android/weatherapp/ui/App.kt]: Yeah, you really only have 1 application context, which is basically just your app.
Nice overview on the different contexts. https://possiblemobile.com/2013/06/context/

File: app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt (hunk 6 above)
Comment (nishtahir) [reply app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt]: Generated code, is not always necessary code. I turn that off mostly because I feel it's better to rely on version control for that info

File: app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt (hunk 8 above)
Comment (nishtahir) [reply app/src/main/java/com/bignerdranch/android/weatherapp/Data/Request.kt]: You instantiate it in your provider methods if you are using dagger.

File: app/src/main/java/com/bignerdranch/android/weatherapp/ui/Adapters/ForecastListAdapter.kt (hunk 11 above)
Comment (nishtahir) [reply app/src/main/java/com/bignerdranch/android/weatherapp/ui/Adapters/ForecastListAdapter.kt]: Looks pretty messy to me. Your view holder should be able to hold references to the views inside. Having to do `itemView.myThing` everywhere is a little repetitive. Another not as clean option IMO is to do 

```
with(itemView) {
   thing.text = "Party parrot"
}
```

File: app/src/main/java/com/bignerdranch/android/weatherapp/Domain/Commands/Command.kt (hunk 9 above)
Comment (nishtahir) [reply app/src/main/java/com/bignerdranch/android/weatherapp/Domain/Commands/Command.kt]: You kind of want the common interface between them right? That's how they know to override `execute` also don't leave comments like this in your code. You want to rely on code review tools for this.
